cloudinit/config/cc_autoinstall.py

```diff
@@ -0,0 +1,140 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+
+"""Autoinstall: Support ubuntu live-server autoinstall syntax."""
+
+import re
+from textwrap import dedent
+
+from cloudinit import log as logging
+from cloudinit.config.schema import (
+    MetaSchema,
+    SchemaValidationError,
+    get_meta_doc,
+)
+from cloudinit.settings import PER_INSTANCE
+from cloudinit.subp import subp
+
+LOG = logging.getLogger(__name__)
+
+distros = ["ubuntu"]
+
+meta: MetaSchema = {
+    "id": "cc_autoinstall",
+    "name": "Autoinstall",
+    "title": "Support Ubuntu live-server install syntax",
+    "description": dedent(
+        """\
+        Ubuntu's autoinstall syntax supports single-system automated installs
+        in either the live-server or live-desktop installers.
+        When "autoinstall" directives are provided in either
+        #cloud-config user-data or ``/etc/cloud/cloud.cfg.d`` validate
+        minimal autoinstall schema adherance and emit a warning if the
+        live-installer is not present.
+
+        The live-installer will use autoinstall directives to seed answers to
+        configuration prompts during system install to allow for a
+        "touchless" Ubuntu system install.
+
+        For more details on Ubuntu's autoinstaller:
+            https://ubuntu.com/server/docs/install/autoinstall
+    """
+    ),
+    "distros": distros,
+    "examples": [
+        dedent(
+            """\
+            # Tell the live-server installer to provide dhcp6 network config
+            # and LVM on a disk matching the serial number prefix CT
+            autoinstall:
+              version: 1
+              network:
+                version: 2
+                ethernets:
+                  enp0s31f6:
+                    dhcp6: yes
+              storage:
+                layout:
+                  name: lvm
+                  match:
+                    serial: CT*
+        """
+        )
+    ],
+    "frequency": PER_INSTANCE,
+}
+
+__doc__ = get_meta_doc(meta)
+
+
+LIVE_INSTALLER_SNAPS = ("subiquity", "ubuntu-desktop-installer")
+
+
+def handle(name, cfg, cloud, log, _args):
+
+    if "autoinstall" not in cfg:
+        LOG.debug(
+            "Skipping module named %s, no 'autoinstall' key in configuration",
+            name,
+        )
+        return
+
+    snap_list, _ = subp(["snap", "list"])
+    installer_present = None
+    for snap_name in LIVE_INSTALLER_SNAPS:
+        if re.search(snap_name, snap_list):
+            installer_present = snap_name
+    if not installer_present:
+        LOG.warning(
+            "Skipping autoinstall module. Expected one of the Ubuntu"
+            " installer snap packages to be present: %s",
+            ", ".join(LIVE_INSTALLER_SNAPS),
+        )
+        return
+    validate_config_schema(cfg["autoinstall"])
```

Comment:
`validate_config_schema` expects the raw config.

```suggestion
    validate_config_schema(cfg)
```